File: src/main/c/jni-lib/src/events/TradeMapping.cpp
// SPDX-License-Identifier: MPL-2.0

#include "dxfeed/events/TradeMapping.hpp"
#include "dxfeed/utils/ByteReader.hpp"
#include "dxfeed/utils/ByteWriter.hpp"

namespace dxfeed::jni {
  dxfg_trade_base_t* TradeMapping::toTradeBase(ByteReader& reader, dxfg_event_clazz_t dxfgEventClazz)
  {
    auto* trade = new dxfg_trade_base_t();
    trade->market_event.event_type.clazz = dxfgEventClazz;
    trade->market_event.event_symbol = reader.readString();
    trade->market_event.event_time = reader.readLong();
    trade->time_sequence = reader.readLong();
    trade->time_nano_part = reader.readInt();
    trade->exchange_code = reader.readInt16_t();
    trade->day_id = reader.readInt();
    trade->flags = reader.readInt();

    trade->price = reader.readDouble();
    trade->change = reader.readDouble();
    trade->size = reader.readDouble();
    trade->day_volume = reader.readDouble();
    trade->day_turnover = reader.readDouble();
    return trade;
  }

  void TradeMapping::fromTradeBase(dxfg_trade_base_t* trade, ByteWriter& writer) {
    writer.writeString(trade->market_event.event_symbol);
    writer.writeInt64_t(trade->market_event.event_time);
    writer.writeInt64_t(trade->time_sequence);
    writer.writeInt32_t(trade->time_nano_part);
    writer.writeInt16_t(trade->exchange_code);
    writer.writeInt32_t(trade->day_id);
    writer.writeInt32_t(trade->flags);

    writer.writeDouble(trade->price);
    writer.writeDouble(trade->change);
    writer.writeDouble(trade->size);
    writer.writeDouble(trade->day_volume);
    writer.writeDouble(trade->day_turnover);
  }
}

File: src/main/java/com/dxfeed/api/serializers/TradeMapping.java
package com.dxfeed.api.serializers;

import com.dxfeed.api.buffers.ByteBuffer;
import com.dxfeed.api.buffers.DoubleBuffer;
import com.dxfeed.api.buffers.NativeEventsReader;
import com.dxfeed.event.market.DxFeedEventMarketPackagePrivate;
import com.dxfeed.event.market.Trade;
import com.dxfeed.event.market.TradeBase;

public class TradeMapping {
  /**
   * typedef struct dxfg_market_event_t {
   *    dxfg_event_type_t event_type;
   *    const char *event_symbol;
   *    int64_t event_time;
   * } dxfg_market_event_t;
   *
   * https://github.com/dxFeed/dxfeed-graal-native-sdk/blob/main/src/main/c/api/dxfg_events.h#L286
   *
   * typedef struct dxfg_trade_base_t {
   *    dxfg_market_event_t market_event;
   *    //    dxfg_lasting_event_t lasting_event;
   *    int64_t time_sequence;
   *    int32_t time_nano_part;
   *    int16_t exchange_code;
   *    double price;
   *    double change;
   *    double size;
   *    int32_t day_id;
   *    double day_volume;
   *    double day_turnover;
   *    int32_t flags;
   * } dxfg_trade_base_t;
   */
  public static void toNative(TradeBase event, ByteBuffer pBytes, DoubleBuffer pDoubles) {
    // BYTE DATA
    pBytes.writeString(event.getEventSymbol());                       // 2 + eventSymbolLength
    pBytes.writeLong(event.getEventTime());                           // 8
    pBytes.writeLong(event.getTimeSequence());                        // 8
    pBytes.writeInt(event.getTimeNanoPart());                         // 4
    pBytes.writeChar(event.getExchangeCode());                        // 2
    pBytes.writeInt(event.getDayId());                                // 4
    pBytes.writeInt(DxFeedEventMarketPackagePrivate.getFlags(event)); // 4

    // DOUBLE DATA
    pDoubles.write(event.getPrice());
    pDoubles.write(event.getChange());
    pDoubles.write(event.getSizeAsDouble());
    pDoubles.write(event.getDayVolumeAsDouble());
    pDoubles.write(event.getDayTurnover());
  }

  public static Trade fromNative(NativeEventsReader reader) {
    Trade trade = new Trade();
    trade.setEventSymbol(reader.readString());
    trade.setEventTime(reader.readLong());
    trade.setTimeSequence(reader.readLong());
    trade.setTimeNanoPart(reader.readInt());
    trade.setExchangeCode(reader.readChar());
    trade.setDayId(reader.readInt());
    DxFeedEventMarketPackagePrivate.setFlags(trade, reader.readInt());

    trade.setPrice(reader.readDouble());
    trade.setChange(reader.readDouble());
    trade.setSizeAsDouble(reader.readDouble());
    trade.setDayVolumeAsDouble(reader.readDouble());
    trade.setDayTurnover(reader.readDouble());
    return trade;
  }
}
